lemur/deployment/service.py

```diff
@@ -1,15 +1,50 @@
+from flask import current_app
 from lemur import database
 
 
-def rotate_certificate(endpoint, new_cert):
+def rotate_certificate(endpoint, old_cert, new_cert):
     """
     Rotates a certificate on a given endpoint.
 
     :param endpoint:
+    :param old_cert:
     :param new_cert:
     :return:
     """
-    # ensure that certificate is available for rotation
-    endpoint.source.plugin.update_endpoint(endpoint, new_cert)
-    endpoint.certificate = new_cert
+    sni_rotation = old_cert in endpoint.sni_certificates
+    if sni_rotation and endpoint.primary_certificate == old_cert:
+        current_app.logger.warn(
+            f"{old_cert.name} attached to endpoint {endpoint.name} as both primary "
+            "and SNI certificate which is likely unnecessary"
+        )
+        _rotate_primary_certificate(endpoint, new_cert)
```

Comment:
Im trying to think through if we could have a potential issue in the case that this logic block was hit? 
Lets say we end up in a case where a single certificate is indeed both the primary AND in the list of SNI's. For this thought experiment lets say that `cert1` is currently on the endpoint and we want to rotate in `cert2`. 
In this case our endpoint will return the following values (before rotation)
```
my_endpoint.primary_certificate = cert1
my_endpoint.certificates_assoc = [cert1]
```

Once we hit this logic path it will call `_rotate_primary_certificate` which will set the primary certificate to be `cert2`. So now, if I understand correctly, our state looks like the following

```
my_endpoint.primary_certificate = cert2
my_endpoint.certificates_assoc = [cert1]
```

Is there any issues that could come up with `cert1` still being associated as an sni_certificate after rotation of the primary?

Reply:
This will not be a problem in AWS as `cert2` would be served in that case. Additionally, `cert1` would be rotated on the next automated scheduled rotation attempt.

Reply:
That being said we _could_ have Lemur rotate both certificates here if we really wanted to.

Reply:
> Additionally, cert1 would be rotated on the next automated scheduled rotation attempt.

This was my assumption but I wanted to confirm. 

> That being said we could have Lemur rotate both certificates here if we really wanted to.

Given your response in the first comment, I don't think  this is needed. Worst case it sounds like the old SNI cert sticks around for another rotation period before being replaced, which we know will be fine.

Reply:
Agreed, if it helps add context by the way, part of my reasoning for initially choosing to only rotate the primary certificate in this case because that's exactly what happens today. Given that I want to contribute this back to the upstream project, maintaining the least surprising behavior is likely to be ideal although what constitutes surprising behavior can be open to interpretation.